---
title: "Exploration of single-copy core genes and genome quality"
author: "Stijn Wittouck"
date: "July 17, 2019"
output:
  html_document:
    depth: 2
    number_sections: yes
    theme: united
    toc: yes
    toc_float: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE)
```

```{r}
library(tidyverse)
library(ggpubr)
```

Input/output directories: 

```{r}
din <- "input_v3"
dout_all <- "results_v3/01_scgs_and_genome_quality"
dout_paper <- "results_v3/paper_species_taxonomy"
parsed <- "parsed_v3"
```

Create output directories: 

```{r}
if (! dir.exists(dout_all)) dir.create(dout_all, recursive = T)
if (! dir.exists(dout_paper)) dir.create(dout_paper, recursive = T)
if (! dir.exists(parsed)) dir.create(parsed, recursive = T)
```

The goal of this document is to explore the candidate and definitive single-copy core genes (SCGs) for the Lactobacillus genus complex (LGC). We will also visualize quality measures of the genomes; those are based on the SCGs. 

A few notes on terminology: 

* In most cases when we use the word "gene", we are referring to an individual gene in one genome, with a specific sequence. However, when we talk about "single-copy core genes", we are actually referring to gene families. 
* When a gene family (e.g. an SCG) is represented by multiple genes in a single genome, we sometimes refer to those genes as "copies". 
* We sometimes refer to (candidate) SCGs as "profiles", because they are represented by profile HMMs. 

# Load data

We need the following information:

* A table with genes that show sequence similarity to candidate SCGs (and thus might belong to the candidate SCG, which is a gene family). This gene table includes the score of the gene to the SCG, the start and end positions in the profile HMM of the alignment and the genome of the gene. This table has already been pre-filtered to only contain the best-scoring candidate SCG per gene. 
* A table of the candidate SCGs (profiles), with their score cutoffs, the number of genomes they were found in in a single copy and the number of genomes they were found in in multiple copies. 
* A table of the genomes with their SCG-based quality measures: completeness and redundancy. 
* A table of the genomes with their statistics from the Genome Taxonomy Database (GTDB) of Parks et al.

```{r}
genes <- read_csv(paste0(din, "/score_table.csv.zip"))
profiles <- read_csv(paste0(din, "/candidate_scg_table.csv"))
genomes <- read_csv(paste0(din, "/genome_table.csv"))
genomes_parks <- read_csv(paste0(din, "/genomes_parks.csv"))
```

# Inspection of hmmer cutoffs

We calculate the length of the gene alignments to the candidate SCGs and add this to the genes table:

```{r}
genes <- mutate(genes, length = hmm_to - hmm_from)
```

We add variable to the genes table that tells us whether a gene is the best-scoring "copy" of a profile in a genome: 

```{r}
genes <- 
  genes %>%
  group_by(profile) %>%
  arrange(- score) %>%
  mutate(is_best_copy = ! duplicated(genome)) %>%
  ungroup()
```

We take a random subset of profiles to explore in some visualizations: 

```{r}
profiles_subset <- sample(profiles$profile, 25)
```

Visualize gene scores and alignment lengths per profile, as well as the profile-specific score cutoffs that were trained: 

```{r}
genes %>%
  filter(profile %in% !! profiles_subset) %>%
  ggplot(aes(x = length, y = score, col = is_best_copy)) +
  geom_point(size = 0.1) + 
  geom_hline(
    data = profiles %>% filter(profile %in% profiles_subset), 
    aes(yintercept = gene_cutoff), col = "grey"
  ) + 
  facet_wrap(~ profile, scales = "free") + 
  scale_color_brewer(palette = "Paired") + 
  theme_bw() 
ggsave(paste0(dout_all, "/example_profiles_scores.png"), units = "cm", width = 30, height = 20)
```

Same plot, but with length-corrected scores:  

```{r}
genes %>%
  filter(profile %in% !! profiles_subset) %>%
  mutate(score_corrected = score / length) %>%
  group_by(profile) %>%
  arrange(- score_corrected) %>%
  mutate(is_best_copy = ! duplicated(genome)) %>%
  ungroup()  %>%
  ggplot(aes(x = length, y = score_corrected, col = is_best_copy)) +
  geom_point(size = 0.1) + 
  facet_wrap(~ profile, scales = "free") + 
  scale_color_brewer(palette = "Paired") + 
  theme_bw() 
ggsave(paste0(dout_all, "/example_profiles_scores_corrected.png"), units = "cm", width = 30, height = 20)
```

Score densities per profile (important for the interpretation: we show a separate density of best-copies and non-best-copies per profile, but both densities are automatically normalized to have the same area-under-the-curve by ggplot!!): 

```{r}
genes %>%
  filter(profile %in% profiles_subset) %>%
  ggplot(aes(x = score, fill = is_best_copy)) +
  geom_density(alpha = 0.5) + 
  geom_vline(
    data = profiles %>% filter(profile %in% profiles_subset), 
    aes(xintercept = gene_cutoff), col = "black"
  ) + 
  facet_wrap(~ profile, scales = "free")
ggsave(paste0(dout_all, "/example_profiles_scores_densities.png"), units = "cm", width = 30, height = 20)
```

Experimental visulization of cumulative number of genes and number of unique genomes as the score increases: 

```{r}
genes %>%
  filter(profile %in% profiles_subset) %>%
  group_by(profile) %>%
  arrange(- score) %>%
  mutate(n_genomes = cumsum(! duplicated(genome))) %>%
  mutate(n_genes = 1:n()) %>%
  gather(value = "count", key = "feature", n_genomes, n_genes) %>%
  ggplot(aes(x = 1 - score, y = count, color = feature)) +
  geom_line() + 
  geom_point(size = 0.1) + 
  facet_wrap(~ profile, scales = "free")
ggsave(paste0(dout_all, "/example_profiles_genes_vs_genomes.png"), units = "cm", width = 30, height = 20)
```

# Inspection of candidate SCGs

For each profile, we have a count of genomes where it occurs in a single copy and a count of genomes where it occurs in multiple copies. We convert these counts to percentages of the total number of genomes: 

```{r}
n_genomes <- genes$genome %>% unique() %>% length()
profiles <-
  profiles %>%
  mutate(
    single_copy = single_copy_presence / !! n_genomes,
    multi_copy = multi_copy_presence / !! n_genomes
  )
```

Visualize single- and multi-copy presence of candidate SCGs: 

```{r}
profiles %>%
  arrange(- single_copy_presence) %>%
  mutate(profile = 1:n()) %>%
  gather(
    value = "percentage_of_genomes", key = "copy_number", single_copy, 
    multi_copy
  ) %>%
  ggplot(aes(
    x = profile, y = percentage_of_genomes, fill = copy_number, 
    col = copy_number
  )) +
  geom_col() + 
  geom_hline(yintercept = 0.95) + 
  geom_vline(
    xintercept = profiles %>% filter(single_copy >= 0.95) %>% nrow(), 
    color = "grey", lty = 2
  ) + 
  scale_fill_brewer(palette = "Paired") + 
  scale_color_brewer(palette = "Paired") + 
  theme_bw() +
  theme(legend.position = "bottom")
ggsave(
  paste0(dout_all, "/profile_selection.png"), units = "cm", width = 17.4, 
  height = 12
)
ggsave(
  paste0(dout_paper, "/figure_S1_profile_selection.pdf"), units = "cm", 
  width = 17.4, height = 12
)
```

Number of profiles that pass the 95% single-copy presence cutoff:

```{r}
passing_profiles <- 
  profiles %>% 
  filter(single_copy >= 0.95) %>% 
  pull(profile)
length(passing_profiles)
```

Score densities for random sample of selected profiles (same important interpretation note as previous density plot!!): 

```{r}
profiles_selected <- sample(passing_profiles, 25)
genes %>%
  filter(profile %in% !! profiles_selected) %>%
  ggplot(aes(x = score, fill = is_best_copy, col = is_best_copy)) +
  geom_density(alpha = 0.5) + 
  geom_rug() + 
  geom_vline(
    data = profiles %>% filter(profile %in% !! profiles_selected), 
    aes(xintercept = gene_cutoff), col = "black"
  ) + 
  facet_wrap(~ profile, scales = "free")
ggsave(paste0(dout_all, "/example_passing_profiles_scores_densities.png"), units = "cm", width = 30, height = 20)
```

# Selection of genomes

Visualize density of the completeness values of the genomes:

```{r}
(
  fig_genomes_completeness <- 
    genomes %>%
    ggplot(aes(x = completeness)) + 
    geom_density() +
    geom_rug() + 
    geom_vline(xintercept = 0.90, color = "black", lty = 2) + 
    xlim(c(0, 1)) + 
    theme_bw() +
    theme(
      axis.ticks.y = element_blank(),
      axis.text.y = element_blank()
    )
)
ggsave(paste0(dout_all, "/genomes_completeness.png"), units = "cm", width = 16, height = 10)
```

Visualize density of the redundancy values of the genomes:

```{r}
(
  fig_genomes_redundancy <- 
    genomes %>%
    ggplot(aes(x = redundancy)) + 
    geom_density() +
    geom_rug() + 
    geom_vline(xintercept = 0.10, color = "black", lty = 2) + 
    xlim(c(0, 1)) + 
    theme_bw() +
    theme(
      axis.ticks.y = element_blank(),
      axis.text.y = element_blank()
    )
)
ggsave(paste0(dout_all, "/genomes_redundancy.png"), units = "cm", width = 16, height = 10)
```

Completeness vs redundancy scatterplot: 

```{r}
genomes %>%
  ggplot(aes(x = completeness, y = redundancy)) + 
  geom_point(size = 1) +
  xlim(c(0, 1)) + ylim(c(0, 1)) + 
  theme_bw()
ggsave(
  paste0(dout_all, "/genomes_completeness_vs_redunancy.png"), 
  units = "cm", width = 16, height = 16
)
```

# Comparison of legen quality control with checkm quality control

Let's prepare a table we can use to compare the completeness and redundancy/contamination of the genomes as estimated by our legen pipeline vs the CheckM tool: 

```{r}
genomes_quality <-
  genomes %>%
  left_join(genomes_parks, by = c("genome" = "accession")) %>%
  select(
    legen_completeness = completeness, legen_redundancy = redundancy,
    checkm_completeness, checkm_contamination, genome
  ) %>%
  mutate_at(vars(contains("checkm")), ~ . / 100) 
```

We can now compare completeness estimates using a scatterplot: 

```{r}
genomes_quality %>%
  ggplot(aes(x = 1 - legen_completeness, y = 1 - checkm_completeness)) +
  geom_jitter(alpha = 0.1) +
  scale_x_log10(limits = c(0.001, 1)) +
  scale_y_log10(limits = c(0.001, 1)) +
  theme_bw()
ggsave(
  paste0(dout_all, "/genomes_checkm_completeness.png"), 
  units = "cm", width = 16, height = 16
)
```

A first observation is that many of the genomes in our dataset, 402 of them, are not represented in the GTDB. For these, we were not able to recover CheckM genome quality measures. For the ones that are represented in GTDB, the correlation between the completeness as estimated by our pipeline and CheckM is there but seems relatively weak. For high completeness values, between 0.99 and 1.00, the correlation seems even nonexistent. Fortunately though, we are of course mainly interested in identifying genomes of poor quality so this isn't a huge problem. 

We can also compare redunancy/contamination (these concepts are similar) estimates: 

```{r}
genomes_quality %>%
  ggplot(aes(x = legen_redundancy, y = checkm_contamination)) +
  geom_jitter(alpha = 0.1) +
  scale_x_log10(limits = c(0.001, 1)) +
  scale_y_log10(limits = c(0.001, 1)) +
  theme_bw()
ggsave(
  paste0(dout_all, "/genomes_checkm_redundancy.png"), 
  units = "cm", width = 16, height = 16
)
```

We can see a similar pattern here: a relatively weak correlation, epecially for low contamination values (under ~ 0.03). 

Let's see what would happen if we select our genomes based on CheckM instead of our own pipeline, using the same cutoffs that we used:

```{r}
genomes_quality %>%
  mutate(
    legen_passes_qc = legen_completeness > 0.90 & legen_redundancy < 0.10,
    checkm_passes_qc = checkm_completeness > 0.90 & checkm_contamination < 0.10
  ) %>%
  count(legen_passes_qc, checkm_passes_qc) %>%
  filter(! is.na(checkm_passes_qc))
```

If we had used CheckM, we would have included 12 genomes that we now excluded, and excluded 15 genomes that we now included. 33 genomes are excluded by both approaches and 2096 included by both approaches. These numbers are of course only for the genomes that had CheckM values available. 

Conclusion: our approach and CheckM clearly yield different results, which is interesting, but the correlation is definitely there and switching to CheckM would not lead to radically different decisions in terms of which genomes to throw away. 

# Species taxonomy paper: compose figure 1

```{r}
give_letter <- function(plot, letter) {
  
  g <- ggplotGrob(plot + ggtitle(letter))
  g$layout$l[g$layout$name == "title"] <- 1
  
  g
  
}

ggarrange(
  fig_genomes_completeness %>% give_letter("A"), 
  fig_genomes_redundancy %>% give_letter("B"), 
  nrow = 2, ncol = 1
)
ggsave(
  paste0(dout_paper, "/figure_1_genome_quality.png"), 
  units = "cm", width = 17.4, height = 10
)
ggsave(
  paste0(dout_paper, "/figure_1_genome_quality.pdf"), 
  units = "cm", width = 17.4, height = 10
)
```