---
title: "Analysis of taxonomy and reclassifications"
author: "Stijn Wittouck"
date: "June 19, 2019"
output:
  html_document:
    depth: 2
    number_sections: yes
    theme: united
    toc: yes
    toc_float: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE)
```

```{r}
library(tidyverse)
library(glue)
library(ggpubr)
```

Input/output directories: 

```{r}
din <- "input_v3"
dout_all <- "results_v3/03_taxonomy"
dout_paper <- "results_v3/paper_species_taxonomy"
parsed <- "parsed_v3"
```

Create output directories: 

```{r}
if (! dir.exists(dout_all)) dir.create(dout_all, recursive = T)
if (! dir.exists(dout_paper)) dir.create(dout_paper, recursive = T)
if (! dir.exists(parsed)) dir.create(parsed, recursive = T)
```

The goal of this document is to compare _de novo_ species (genome clusters) to published species and subspecies. The result will be species names for all genome clusters (where possible), suggestions for species mergers/splits and the identification of novel species. We will also classify genomes (strains) that are unclassified (on the species level) on NCBI and we will identify genomes that are misclassified. 

# Import data

We need the following information:

* A table with the clusters to which all genomes belong. 
* A table with for each cluster the minimum cni (just extra information to add to table 1). 
* A table with NCBI genome assembly reports for all genomes to know NCBI species labels. 
* A table with validly published names found by tidytypes, to get an idea of how many species and names (including subspecies) have been found.
* For the type genomes approach: a table with type genomes in our data, identified in an automated manner using the R package tidytypes.
* For the type 16S rRNA approach: a table with hits of 16S sequences extracted from the genomes, a table with rRNA gene names of all extracted 16S rRNAs to know which genomes yielded 16S rRNAs and which genomes didn't. 

Let's load all these files:

```{r}
genomes_clusters <- 
  paste0(din, "/genomes_clusters.csv") %>%
  read_csv()
clusters_exlusivity <- 
  paste0(parsed, "/clusters_exclusivity.csv") %>%
  read_csv() 
genomes_ncbi <- 
  paste0(din, "/genomes_assembly_reports.csv") %>%
  read_csv() 
load(paste0(din, "/lgc_names.rda"))
genomes_type_automatic <- 
  paste0(din, "/lgc_genomes_type.csv") %>%
  read_csv() %>%
  left_join(genomes_clusters)
sixteen_s_hits <- 
  paste0(din, "/hits_to_type_16s_genes.tsv") %>%
  read_tsv(col_names = F) %>%
  set_names(c("qaccver", "saccver", "pident", "length", 
      "mismatch", "gapopen", "qstart", "qend", 
      "sstart", "send", "evalue", "bitscore"
  )) %>%
  mutate(genome = str_extract(qaccver, "^[^:]+"))
rrnas <- 
  paste0(din, "/rrnas_16S.txt") %>%
  read_lines()
```

Quick overview of valid names:

```{r}
glue("number of names: {nrow(names_combined)}")
glue("number of species: {unique(names_combined$species) %>% length()}")
```

Quick overview of automatically identified type genomes, as well a the number of names (species or subspecies) and the number of species they belong to:

```{r}
genomes_type_automatic %>% nrow() 
genomes_type_automatic$name %>% unique() %>% length() 
genomes_type_automatic$species %>% unique() %>% length() 
```

Finally, we'll create a general function to convert full species names to abbreviated names, to use for tables and figures. 

```{r}
abbreviate <- function(species) {
  species %>%
    str_replace_all("Lactobacillus", "L.") %>%
    str_replace_all("Pediococcus", "P.") %>%
    str_replace_all("Leuconostoc", "Leuc.") %>%
    str_replace_all("Oenococcus", "O.") %>%
    str_replace_all("Fructobacillus", "F.") %>%
    str_replace_all("Weissella", "W.") %>%
    str_replace_all("Convivina", "C.") %>%
    str_replace_all("Unidentified", "Unident.") %>%
    str_replace_all("pseudomesenteroides", "pseudomesent.")
}
```


# Preparation of type genomes

## Manual identification of extra type genomes

What species names can we find on NCBI that we didn't yet identify a type genome for using the automated approach?

```{r}
genomes_ncbi$species %>% 
  unique() %>%
  str_extract("^[^ ]+ [^ ]+") %>%
  keep(~ ! str_detect(., "sp\\.| bacterium")) %>%
  keep(~ ! . %in% genomes_type_automatic$species) %>%
  sort()
```

Let's see which of those species names actually do have type genomes on NCBI and list the reason why they weren't identified automatically:

```{r}
genomes_type_manual <-
  tribble(
    ~ name, ~ genome, ~ why_manual,
    "Convivina intestini", "GCF_003096575.1", "separate genus",
    "Lactobacillus ginsenosidimutans", "GCF_001050475.1", "effectively published",
    "Lactobacillus kosoi", "GCF_003112615.1", "effectively published",
    "Lactobacillus musae", "GCF_003573575.1", "strain name only numbers",
    "Lactobacillus paragasseri", "GCF_003584685.1", "very recent",
    "Lactobacillus raoultii", "GCF_900411375.1", "effectively published", 
    "Lactobacillus terrae", "GCF_002762335.1", "very recent", 
    "Lactobacillus timonensis", "GCF_900240275.1", "effectively published",
    "Lactobacillus nuruki", "GCF_002872255.1", "genus name wrongly spelled",
    "Leuconostoc lactis", "GCF_000179875.1", "synonym type material",
    "Leuconostoc suionicum", "GCF_001891125.1", "more recent publication"
  ) %>%
  left_join(genomes_clusters) %>%
  mutate(species = str_extract(name, "^[^ ]+ [^ ]+"))
```

For the other species names: why are NCBI genomes classified to those species if the type strain of the species is not sequenced (and of high quality)? 

* Lactobacillus backii: NCBI genomes are not type (checked pubmed); not clear why/how they were classified to this species. 
* Lactobacillus bombi: NCBI genomes are not type (checked pubmed); three genomes from the same study; not clear why/how they were classified to this species. 
* Lactobacillus coleohominis: type genome on NCBI but its assembly is too poor quality
* Lactobacillus kefiri: type genome on NCBI but its assembly is too poor quality (two type genomes even)
* Lactobacillus rogosae: type genome on NCBI (GCF_900112995.1) but its assembly is too poor quality
* Leuconostoc carnosum: type genome on NCBI but its assembly is too poor quality
* Leuconostoc citreum: NCBI genomes are not type (checked LPSN, including heterotypic synonym Leuconostoc amelibiosum); not clear why/how they were classified to this species. 
* Leuconostoc garlicum: not on LPSN, no original paper found that describes the species; not clear why/how they were classified to this species. 
* Leuconostoc pseudomesenteroides: type genome on NCBI but its assembly is too poor quality
* Oenococcus alcoholitolerans: type genome on NCBI (GCA_000769695.1) but its assembly is too poor quality
* Pediococcus parvulus: NCBI genomes are not type (checked pubmed); not clear why/how they were classified to this species. 
* Sharpea azabuensis: type genome on NCBI (GCA_000702165.1) but its assembly is too poor quality; according to the GTDB, the Sharpea genus doesn't belong to the Lactobacillales
* Weissella bombi: NCBI genomes are not type (checked pubmed); not clear why/how they were classified to this species. 
* Weissella ceti: NCBI genomes are not type (checked pubmed); not clear why/how they were classified to this species. 
* Weissella cibaria: validly published; strain KACC 11862 has a genome on NCBI that passes our QC and in a paper this strain is called the type strain of the species, but the paper does not refer to the "established" (LPSN) names of the type strain, so untrustworthy. 
* Weissella hellenica: NCBI genomes are not type (checked LPSN); not clear why/how they were classified to this species. 
* Weissella koreensis: type genome on NCBI but its assembly is too poor quality

Let's combine the automatically and manually identified type genomes:

```{r}
genomes_type <- bind_rows(genomes_type_automatic, genomes_type_manual)
```

## Remove wrong type genomes

Further in the pipeline, we noticed that all genomes of Weissella cibaria are reclassified to Leuconostoc miyukkimchii. This is due to the fact that someone uploaded a genome of the former species and gave it the strain name "M2", which is coincidentally also one of the names for the Leuconostoc miyukkichmii type strain. We remove this genome from the type genomes: 

```{r}
genomes_type <-
  genomes_type %>%
  anti_join(tribble(
    ~ strain_name, ~ species,
    "M2", "Leuconostoc miyukkimchii"
  ))
```

Inspect cases where the same genome is type for different names: 

```{r}
genomes_type %>%
  select(genome, name) %>%
  add_count(genome) %>%
  filter(n >= 2)
```

L. suionicum is the more recent name, so let's pick that one:

```{r}
genomes_type <-
  genomes_type %>%
  anti_join(tribble(
    ~ genome, ~ species,
    "GCF_001891125.1", "Leuconostoc mesenteroides"
  ))
```

Inspect cases where exactly the same name is in different genome clusters:

```{r}
genomes_type %>%
  group_by(name) %>%
  mutate(n = cluster %>% unique %>% length) %>%
  ungroup() %>%
  filter(n >= 2) %>%
  arrange(name) %>%
  left_join(genomes_ncbi %>% select(genome, species_ncbi = species)) %>%
  select(genome, strain_name, name, cluster, species_ncbi)
```

How can we resolve those inconsistencies?

* Three genomes on NCBI have a strain name beloning to the type strain of Lactobacillus casei (see table). One of them belongs to a different cluster than the other two; we assume that this genome was given an incorrect strain name. 
* Assembly GCF_000965255.1 received strain name LP38, which is a type strain name for Lactobacillus hammesii but the genome is classified to Lactobacillus spicheri on NCBI. The type strain name match is probably coincidental. 
* Assembly GCF_000091765.1 received strain name ST1, which is a type strain name for Lactobacillus panis but the genome is classified to Lactobacillus crispatus on NCBI. The type strain name match is probably coincidental. 
* Assembly GCF_000309665.1 received strain name M36, which is a type strain name for Lactobacillus similis but the genome is classified to Lactobacillus paracasei on NCBI. The type strain name match is probably coincidental. 

Now we can remove those wrong genome - type strain connections:

```{r}
genomes_type <-
  genomes_type %>%
  anti_join(tribble(
    ~ genome, ~ name,
    "GCF_001433735.1", "Lactobacillus casei subsp. casei",
    "GCF_000965255.1", "Lactobacillus hammesii",
    "GCF_000091765.1", "Lactobacillus panis",
    "GCF_000309665.1", "Lactobacillus similis"
  ))
```

Now we can make a table with "combinations" of genome clusters and names:

```{r}
combinations <- 
  genomes_type %>%
  distinct(cluster, species, name) 
```

# Name genome clusters using type genomes

## Species mergers

Clusters with type genomes from different species: 

```{r}
mergers <-
  combinations %>%
  group_by(cluster) %>%
  mutate(n = species %>% unique %>% length) %>%
  ungroup() %>%
  filter(n > 1) %>%
  arrange(cluster)
mergers %>%
  select(cluster, n, name, species) %>%
  print(n = 30)
```

For the mergers, we pick the species name that was established the earliest and remove the later names, by checking the LPSN:

* cluster 135: we keep Lactobacillus amylophilus (sp. nov., 1981) and drop Lactobacillus amylotrophicus (sp. nov., 2006) because the former is older. 
* cluster 141: we randomly drop Lactobacillus bobalius, but actually both those names have been renamed to Lactobacillus paralimentarius.
* cluster 173: we keep Lactobacillus timberlakei (2018) and drop Lactobacillus micheneri (2018) purely randomly; they are described in the same publication. We also drop Lactobacillus kosoi (2018) because it is only effectively published. 
* cluster 178: we keep Weissella thailandensis (2000) and drop Weissella jogaejeotgali (2015) because the former is older. 
* cluster 179: we keep Lactobacillus fructivorans (described in 1934) and drop Lactobacillus homochiochii (1957), because the former is listed as an earlier heterotypic synonym of the latter. 
* cluster 218: we keep Pediococcus acidilactii (described in 1887) and drop Pediococcus lolii (2006), because the former is older. On LPSN, the question is already raised if the two species names are heterotypic synonyms. Also, it is mentioned that the P. lolii type strain is not available from culture collections. 
* cluster 225: we keep Leuconostoc gelidum (1989) and drop Leuconostoc gasicomitatum (2001) and Leuconstoc inhae (2003), because the former is oldest. 
* cluster 233: we keep Lactobacillus gasseri (1980) and drop Lactobacillus paragasseri (2018) because the former is older.
* cluster 236: we keep Leuconostoc mesenteroides (1878) and drop Leuconostoc suionicum (2017) because the former is older. 
* cluster 238: we keep Lactobacillus casei and drop Lactobacillus zeae, based on [this page on the casei/paracasei/zeae issue](http://www.bacterio.net/notecasei.html).

```{r}
species_to_drop <- 
  c(
    "Lactobacillus amylotrophicus",
    "Lactobacillus bobalius",
    "Lactobacillus micheneri",
    "Lactobacillus kosoi",
    "Weissella jogaejeotgali",
    "Lactobacillus homohiochii",
    "Pediococcus lolii",
    "Leuconostoc gasicomitatum",
    "Leuconostoc inhae",
    "Lactobacillus paragasseri",
    "Leuconostoc suionicum",
    "Lactobacillus zeae"
  )
combinations2 <- 
  combinations %>%
  filter(! species %in% species_to_drop)
```

## Species splits

Species with type genomes in multiple clusters: 

```{r}
splits <-
  combinations %>%
  group_by(species) %>%
  mutate(n = cluster %>% unique %>% length) %>%
  ungroup() %>%
  filter(n > 1) %>%
  arrange(species) 
splits
```

We suggest an upgrade of L. aviarius subsp. araffinosus to L. araffinosus. 

```{r}
combinations3 <-
  combinations2 %>%
  mutate(name = if_else(name == "Lactobacillus aviarius subsp. araffinosus", "Lactobacillus araffinosus", name)) %>%
  mutate(species = if_else(name == "Lactobacillus araffinosus", "Lactobacillus araffinosus", species))
```

Now for our ultimate result of this part: a table with clusters named using automatically and manually identified type genomes (tgs): 

```{r}
clusters_tgs <-
  combinations3 %>%
  distinct(cluster, species) 
```

We also write out table 1, which is an overview of proposed species splits and mergers:

```{r}
splits_and_mergers <-
  bind_rows(
    mergers %>% mutate(merger_split = "merger"),
    splits %>% mutate(merger_split = "split")
  ) %>%
  group_by(cluster, merger_split) %>%
  summarize(names = str_c(name, collapse = ", ")) %>%
  left_join(clusters_exlusivity) %>%
  select(cluster, merger_split, names, min_cni_within) %>%
  mutate_at("names", abbreviate) %>%
  arrange(merger_split, desc(min_cni_within))
splits_and_mergers %>% write_csv(paste0(dout_all, "/splits_and_mergers.csv"))
splits_and_mergers %>% write_csv(paste0(dout_paper, "/table_1_splits_and_mergers.csv"))
```

# Name genome clusters using 16S rRNA hits and NCBI labels

Let's make an overview of clusters not named used the type genomes approach; we call them "type 16S sequence (tss) clusters" because we will attempt to name them using 16S rRNA sequences of type species: 

```{r}
clusters_tss <-
  genomes_clusters %>%
  filter(! cluster %in% clusters_tgs$cluster) %>%
  left_join(genomes_ncbi) %>%
  select(cluster, species_ncbi = species) %>%
  group_by(cluster) %>%
  summarize(
    species_ncbi_all = 
      species_ncbi %>% 
      unique() %>%
      keep(~ ! str_detect(., "sp\\.| bacterium ")) %>%
      str_c(collapse = ", ") %>%
      when(length(.) == 0 ~ as.character(NA), ~ .),
    species_ncbi = 
      species_ncbi %>% 
      unique() %>%
      keep(~ ! str_detect(., "sp\\.| bacterium ")) %>%
      keep(~ ! . %in% clusters_tgs$species) %>% 
      str_c(collapse = ", ") %>%
      when(length(.) == 0 ~ as.character(NA), ~ .),
    n_genomes =
      n()
  ) %>%
  arrange(species_ncbi)
```

## Process comparison to 16S rRNA sequences 

Candidate names of clusters based on 16S hits:

```{r}
clusters_candidate_names_16S <-
  sixteen_s_hits %>%
  left_join(genomes_clusters) %>%
  filter(pident >= 98, length >= 100) %>%
  rename(sixteen_s_hit = saccver) %>%
  distinct(cluster, sixteen_s_hit) %>%
  group_by(cluster) %>%
  summarize(
    sixteen_s_hits = str_c(sixteen_s_hit, collapse = ", ")
  )
```

Numbers of 16S genes found per cluster:

```{r}
clusters_n_16S <-
  rrnas %>%
  tibble(genome = .) %>%
  left_join(genomes_clusters) %>%
  count(cluster) %>%
  rename(n_16S = n)
```

Overview of all info we have on unnamed clusters:

```{r}
clusters_tss <-
  clusters_tss %>%
  left_join(clusters_candidate_names_16S) %>%
  left_join(clusters_n_16S) %>%
  replace_na(list(n_16S = 0)) %>%
  mutate(
    status = 
      case_when(
        n_16S == 0 ~ "no 16S sequences",
        map_lgl(sixteen_s_hits, is.na) ~ "16S sequences without hits",
        TRUE ~ "16S sequences with hits"
      )
  ) 
```

## Further research the genomes in the unnamed clusters

For a number of these clusters, we could find exactly one species label on NCBI that we didn't already use to name a cluster. In all of those cases, the NCBI species label was also found using the 16S blast approach. We will assume that those species labels are indeed correct, but we will add a question mark to indicate that they are not 100% certain (since they are not based on similarity to type genomes). 

```{r}
clusters_tss <-
  clusters_tss %>%
  mutate(
    certainty = case_when(
      ! is.na(species_ncbi) & status == "16S sequences with hits" ~ "best guess",
      status == "16S sequences without hits" ~ "new species",
      TRUE ~ "unidentified"
    )
  ) %>%
  group_by(certainty) %>%
  mutate(
    species = case_when(
      certainty == "best guess" ~ str_c(species_ncbi, " (?)"),
      certainty == "new species" ~ str_c("New species", 1:n(), sep = " "),
      certainty == "unidentified" ~ str_c("Unidentified species", 1:n(), sep = " ")
    )
  ) %>%
  ungroup()
```

Let's write table 2; an overview of clusters we could not name using type genomes: 

```{r}
clusters_zerotypegenomes <-
  clusters_tss %>%
  mutate_at(
    "certainty", factor, 
    levels = c("best guess", "new species", "unidentified")
  ) %>%
  arrange(certainty)  %>%
  select(cluster, species_ncbi, n_16S, sixteen_s_hits, species) %>%
  mutate_all(str_replace_all, "_", " ") %>%
  mutate_all(abbreviate)
clusters_zerotypegenomes %>% write_csv(paste0(dout_all, "/clusters_zerotypegenomes.csv"))
clusters_zerotypegenomes %>% write_csv(paste0(dout_paper, "/table_2_clusters_zerotypegenomes.csv"))
```

Now to make a table with all cluster names that we found/chose:

```{r}
clusters_all <-
  bind_rows(
    clusters_tgs,
    clusters_tss %>% select(cluster, species)
  ) %>%
  mutate(species_short = abbreviate(species))
write_csv(clusters_all, paste0(parsed, "/clusters_all_named.csv"))
```

# Visualization of reclassifications

## Visualize classification of unclassified

Make a genome table with all genomes and their NCBI and CNI species label:

```{r}
genomes_taxonomy <-
  left_join(
    genomes_clusters %>%
      left_join(clusters_all) %>%
      select(genome, species_cni = species),
    genomes_ncbi %>%
      select(genome, species_ncbi = species)
  )
```

We make a distinction between genomes that are classified and unclassified on NCBI:

```{r}
unclassified_tags <-
  c("sp\\.", " bacterium") %>%
  str_c(collapse = "|")
genomes_taxonomy <-
  genomes_taxonomy %>%
  mutate(classified_ncbi = ! (str_detect(species_ncbi, !! unclassified_tags) | is.na(species_ncbi))) 
genomes_taxonomy %>%
  count(classified_ncbi)
```

We also correct some NCBI species names that contain extra information:

```{r}
genomes_taxonomy <-
  genomes_taxonomy %>%
  mutate_at("species_ncbi", str_extract, "^[^ ]+ [^ ]+")
```

Bar chart: 

```{r}
(
  fig_reclassifications_unclassified <- 
    genomes_taxonomy %>%
    filter(! classified_ncbi) %>%
    count(species_cni) %>%
    mutate_at("species_cni", abbreviate) %>%
    mutate(species_cni = fct_reorder(species_cni, n) %>% fct_rev()) %>%
    mutate(is_new_species = str_detect(species_cni, "species")) %>%
    mutate_at("is_new_species", if_else, "unknown", "known") %>% 
    ggplot(aes(x = species_cni, y = n)) +
    geom_col() + 
    facet_grid(~ is_new_species, scales = "free_x", space = "free_x") + 
    scale_y_continuous(breaks = function(limits) seq((limits[1] %/% 2) * 2, floor(limits[2]), by = 2)) +
    xlab("genome cluster") +
    ylab("number of genomes") + 
    theme_bw() +
    theme(
      panel.grid.major = element_line(colour = "grey80"),
      text = element_text(size = 8),
      axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
      plot.margin = unit(c(0, 1, 0, 2), "cm")
    )
)
ggsave(paste0(dout_all, "/reclassifications_unclassified.png"), units = "cm", width = 17.4, height = 12)
```

## Visualize reclassifications of genomes that are NOT unclassified in NCBI

Make a genome table with all classified genomes and their SCNI and NCBI species annotation: 

```{r}
genomes_taxonomy_classified <- 
  genomes_taxonomy %>%
  filter(classified_ncbi)
```

Make a list of "boring species" that contain identical genomes in the SCNI and NCBI classifications: 

```{r}
species_cni <- 
  genomes_taxonomy_classified %>%
  count(species_cni, species_ncbi) %>%
  group_by(species_cni) %>%
  summarize(
    inv_simpson_cni = 
      1 / sum((n / sum(n)) ^ 2), n_genomes_cni = sum(n)
  ) 

species_ncbi <- 
  genomes_taxonomy_classified %>%
  count(species_cni, species_ncbi) %>%
  group_by(species_ncbi) %>%
  summarize(
    inv_simpson_ncbi = 
      1 / sum((n / sum(n)) ^ 2), n_genomes_ncbi = sum(n)
  )

species_cni_ncbi <- 
  full_join(
    species_cni %>% rename(species = species_cni),
    species_ncbi %>% rename(species = species_ncbi)
  )

boring_species <- species_cni_ncbi %>%
  filter(inv_simpson_cni == 1, inv_simpson_ncbi == 1) %>%
  pull(species)
```

Compile a table of all genomes belonging to species that are not boring, in either classification: 

```{r}
genomes_to_plot <- 
  genomes_taxonomy_classified %>%
  filter(! species_ncbi %in% !! boring_species) %>%
  mutate(is_new_species = str_detect(species_cni, "species")) %>%
  mutate(species_cni_core = str_extract(species_cni, "^[^ ]+ [^ ]+")) %>%
  mutate(reclassified = if_else(species_cni_core != species_ncbi, "yes", "no")) %>%
  select(genome, species_cni, species_ncbi, reclassified, is_new_species)
```

How many genomes do we reclassify? 

```{r}
count(genomes_to_plot, reclassified)
```

What are the most frequent reclassifications, and what type of reclassification are they?

```{r}
reclassification_counts <-
  genomes_to_plot %>%
  mutate(
    type_of_reclassification = case_when(
      species_ncbi %in% !! species_to_drop ~ "due to merger",
      species_cni == "Lactobacillus araffinosus" ~ "due to split", 
      str_detect(species_cni, "[0-9]") ~ "new/unidentified species",
      TRUE ~ "correction of misclassification"
    )
  ) %>%
  filter(reclassified == "yes") %>%
  count(species_ncbi, species_cni, type_of_reclassification)
reclassification_counts %>%
  arrange(- n) %>%
  print(n = 30)
```

How much reclassifications of each type did we perform? 

```{r}
reclassification_counts %>%
  group_by(type_of_reclassification) %>%
  summarize(n = sum(n)) 
```

Scatterplot:

```{r}
(
  fig_reclassifications_classified <-
    genomes_to_plot %>%
    mutate_at("species_cni", abbreviate) %>%
    mutate_at("species_ncbi", abbreviate) %>%
    mutate(species_cni = fct_reorder(species_cni, is_new_species)) %>%
    count(species_cni, species_ncbi, reclassified) %>%
    rename(n_genomes = n) %>%
    ggplot(aes(x = species_ncbi, y = species_cni, col = reclassified, size = n_genomes)) +
    geom_point() +
    scale_color_manual(values = c("yes" = "#e31a1c", "no" = "#1f77b4")) + 
    scale_size(name = "number of genomes") +
    xlab("NCBI species") + ylab("genome cluster") + 
    theme_bw() +
    theme(
      panel.grid.major = element_line(colour = "grey80"),
      text = element_text(size = 8),
      axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
      legend.position = "bottom"
    )
)
ggsave(paste0(dout_all, "/reclassifications_classified.png"), units = "cm", width = 17.4, height = 18)
```

# Species taxonomy paper: compose figure 4

Load necessary libraries to make figure panel and custom function to given figure letters: 

```{r}
give_letter <- function(plot, letter) {
  
  g <- ggplotGrob(plot + ggtitle(letter))
  g$layout$l[g$layout$name == "title"] <- 1
  
  g
  
}
```

Panel:

```{r}
ggarrange(
  fig_reclassifications_unclassified %>% give_letter("A"), 
  fig_reclassifications_classified %>% give_letter("B"),
  ncol = 1, nrow = 2, heights = c(1, 2.5)
)
ggsave(paste0(dout_paper, "/figure_4_reclassifications.eps"), units = "cm", width = 17.4, height = 23)
```
